x/lazyledgerapp/client/builder.go

```diff
@@ -0,0 +1,172 @@
+	chainID              string
+	accountNum, sequence uint64
+}
+
+func NewBuilder(address sdktypes.Address, chainID string) *Builder {
+	encCfg := params.RegisterAccountInterface(params.MakeEncodingConfig())
+	txBuilder := encCfg.TxConfig.NewTxBuilder()
+	return &Builder{
+		TxBuilder: txBuilder,
+		encCfg:    encCfg,
+		address:   address,
+		chainID:   chainID,
+	}
+}
+
+// UpdateAccountNumber queries the applicaiton to find the latest account number and
+// sequence, updating the respective internal fields
+func (b *Builder) UpdateAccountNumber(ctx context.Context, conn *grpc.ClientConn) error {
+	accNum, seqNumb, err := QueryAccount(ctx, conn, b.encCfg, b.address.String())
+	if err != nil {
+		return err
+	}
+	b.accountNum = accNum
+	b.sequence = seqNumb
+	return nil
+}
+
+// BuildSignedTx creates a signed sdk.Tx that contains the provided WirePayForMessage.
+// The internal fees, the account number, and the sequence should be updated before calling this function
+// (see b.UpdateAccountNumber, sdkclient.TxBuilder.SetGasLimit, and sdkclient.TxBuilder.SetFeeAmount)
+func (b *Builder) BuildSignedTx(msg sdktypes.Msg, kr keyring.Keyring) (authsigning.Tx, error) {
+	// set the msg
+	err := b.SetMsgs(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	// lookup account info
+	keyInfo, err := kr.KeyByAddress(b.address)
+	if err != nil {
+		return nil, err
+	}
+
+	// we must first set an empty signature in order generate
+	// the correct sign bytes
+	sigV2 := signing.SignatureV2{
+		PubKey: keyInfo.GetPubKey(),
+		Data: &signing.SingleSignatureData{
+			SignMode:  signing.SignMode_SIGN_MODE_DIRECT,
+			Signature: nil,
+		},
+		Sequence: b.sequence,
+	}
+
+	// set the empty signature
+	err = b.SetSignatures(sigV2)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate the new signing data
+	signerData := authsigning.SignerData{
+		ChainID:       b.chainID,
+		AccountNumber: b.accountNum,
+		Sequence:      b.sequence,
+	}
+
+	// Generate the bytes to be signed.
+	bytesToSign, err := b.encCfg.TxConfig.SignModeHandler().GetSignBytes(
+		signing.SignMode_SIGN_MODE_DIRECT,
+		signerData,
+		b.GetTx(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sign those bytes using the keyring. we are ignoring the returned public key
+	sigBytes, _, err := kr.SignByAddress(b.address, bytesToSign)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the SignatureV2 struct, this time including a real signature
+	sigV2 = signing.SignatureV2{
+		PubKey: keyInfo.GetPubKey(),
+		Data: &signing.SingleSignatureData{
+			SignMode:  signing.SignMode_SIGN_MODE_DIRECT,
+			Signature: sigBytes,
+		},
+		Sequence: b.sequence,
+	}
+
+	// set the final signature
+	err = b.SetSignatures(sigV2)
+	if err != nil {
+		return nil, err
+	}
+
+	// return the signed transaction
+	return b.GetTx(), nil
+}
+
+// EncodeTx uses the builder's encoding config to encode the provided sdk transaction
+func (b Builder) EncodeTx(tx sdktypes.Tx) ([]byte, error) {
```

Comment:
Previously you used `*Builder` as func receiver, and from now, you're using `Builder` (no asterisk). Is there any reason for this?

Reply:
Not specifically, no. I was only using the general rule of trying not using a pointer unless necessary. I'm happy to change this if you think it would make more sense to keep everything a pointer or for any other reason.